graph/path/earliest.go

```diff
@@ -0,0 +1,100 @@
+package path
+
+import (
+	"gonum.org/v1/gonum/graph"
+	"log"
+	"gonum.org/v1/gonum/graph/temporal"
+)
+
+type Earliest struct {
+	from  graph.Node
+	at    uint64
+	until uint64
+	nodes map[int64]struct{
+		earliest uint64
+		via      []int64
+	}
+}
+
+func (e *Earliest) set(v graph.Node, t uint64, p []int64) {
+	e.nodes[v.ID()] = struct{
+		earliest uint64
+		via      []int64
+	}{
+		t,
+		p,
+	}
+}
+
+func (e *Earliest) From() graph.Node { return e.from }
+
+func (e *Earliest) At() uint64 { return e.at }
+
+func (e *Earliest) Until() uint64 { return e.until }
+
+func (e *Earliest) To(u interface{}) (path []graph.Node, duration uint64) {
+	var uid int64
+	switch t := u.(type) {
+	case graph.Node:
+		uid = t.ID()
+	case int64:
+		uid = t
+	default:
+		log.Panicln("not supported")
+	}
+	if eu, ok := e.nodes[uid]; !ok {
+		return nil, temporal.MaxTime
+	} else {
+		duration = eu.earliest
+		for _, vid := range eu.via {
+			v := temporal.Node(vid)
+			path = append(path, &v)
+		}
+		u := temporal.Node(uid)
+		path = append(path, &u)
+		return
+	}
+}
+
+func (e *Earliest) Len() int {
+	return len(e.nodes)
+}
+
+// EarliestArrivalFrom computes the earliest-arrival paths to all nodes reachable from u where
+// from <= t <=until. It is an implementation of Algorithm 1. described in
+// https://www.vldb.org/pvldb/vol7/p721-wu.pdf
+func EarliestArrivalFrom(g graph.TemporalStream, from graph.Node, at uint64, until uint64) Earliest {
+	earliest := Earliest{
+		from:  from,
+		at:    at,
+		until: until,
+		nodes: make(map[int64]struct{
+			earliest uint64
+			via      []int64
+		}),
+	}
+	earliest.set(from, at, []int64{})
+	s := g.LineStream()
+	for s.Next() {
+		l := s.TemporalLine()
+		u := l.From()
+		uid := u.ID()
+		eu, ok := earliest.nodes[uid]
+		tl := l.At()
+		dtl := l.Duration()
+		if !ok {
```

Comment:
This should happen prior to the calls to `At` and `Duration`; they are discarded if `!ok`.